Improved the type checking for uses of scan.

gnecula opened the pull request:
Previous way of checking was done after flattening and got
easily confused by tuples of different shapes, or None.

Relates to https://github.com/google/jax/issues/1534.

gnecula committed:
Improved the type checking for uses of scan.

Previous way of checking was done after flattening and got
easily confused by tuples of different shapes, or None.

Relates to https://github.com/google/jax/issues/1534.

gnecula committed:
Fixed newly introduced bug.

Also cleaned up deprecation warning in test.

mattjj reviewed (approved):
LGTM! These checks are much better, and thanks for adding the tests for error messages.

mattjj commented on jax/lax/lax_control_flow.py:
```diff
@@ -474,10 +476,16 @@ def scan(f, init, xs):
+  if not isinstance(out_tree_avals, tuple) or len(out_tree_avals) != 2:
+    msg = "scan body output must be a pair, got {}."
+    raise TypeError(msg.format(out_tree_avals))
+  if out_tree_avals[0] != init_tree_avals:
```
While our standard pytrees (`list`/`tuple`/`dict`) define equality, this line would also require user-defined pytree types to define equality. That may be a new constraint on pytrees. I'm fine with adding that constraint here, but an alternative would be to compare the treedefs and separately compare the flattened lists of avals.

gnecula replied:
Good point, I had tried that but I was not able to extract the treedef for the first element of a pair given the treedef for the pair. I need this to check that the returned carry (first element) matches the initial carry.

mattjj replied:
I haven't looked at the details but [`treedef_children`](https://github.com/google/jax/blob/5b724e0019243cc5ebb71d65963e1d829f65999e/jax/tree_util.py#L122) might help us here.

gnecula replied:
Implemented suggestion. Custom pytrees can still be tricky, if the custom flattener stores aux_data that does not support equality.

shoyer commented:
does this actually fix the error reporting problem from #1534, or is it just related?

gnecula committed:
Alternative implementation of type checking for scan.

This avoids comparing pytree of abstract values, because whem using
custom pytrees such things may not be defined and may not support equality.

Instead we compare separately the pytreedef and the abstract values. Note
that this requires pytreedef to be comparable, which in turn requires
that the custom data that a custom pytree may use to be comparable.

gnecula commented:
I have changed the implementation to compare pytreedefs instead of pytrees themselves. It was a good opportunity to learn more about how pytrees work. 
I have also expanded the tests to show the complete actual error message.

gnecula commented:
> does this actually fix the error reporting problem from #1534, or is it just related?

It does fix it; one of the test cases directly targets this bug

gnecula committed:
Fix x64 tests
